File: src/preference.pl
new_require('food type', X) :- new_prefer(X).
new_require('establishment', 'shop') :- new_prefer('drink'), not not_require('establishment', 'shop').
new_prefer('bar') :- new_require('food type', 'bar').
new_require('establishment', 'bar') :- new_prefer('bar').
new_require('establishment', 'pub') :- new_prefer('bar'), not not_require('establishment', 'pub').
new_require('establishment', 'sports bar') :- new_prefer('bar'), not not_require('establishment', 'sports bar').
new_require('establishment', 'bar&grill') :- new_prefer('bar'), not not_require('establishment', 'bar&grill').
new_prefer('bar') :- new_prefer('alcohol').
new_prefer('bar') :- new_prefer('drink').
new_require('establishment', 'buffet') :- new_require('establishment', 'restaurant'), not not_require('establishment', 'buffet').

new_not_require('food type', X) :- new_not_prefer(X).
new_not_require('establishment', 'shop') :- new_not_prefer('drink'), not require('establishment', 'shop').
new_not_require('establishment', 'pub') :- new_not_prefer('bar'), not require('establishment', 'pub').
new_not_require('establishment', 'sports bar') :- new_not_prefer('bar'), not require('establishment', 'sports bar').
new_not_require('establishment', 'bar&grill') :- new_not_prefer('bar'), not require('establishment', 'bar&grill').
new_not_prefer('bar') :- new_not_prefer('alcohol').

entail(prefer(X), require('food type', X)).
entail(prefer('bar'), require('establishment', 'pub')).
entail(prefer('bar'), require('establishment', 'sports bar')).
entail(prefer('bar'), require('establishment', 'bar&grill')).
entail(prefer('alcohol'), require('establishment', 'pub')).
entail(prefer('alcohol'), require('establishment', 'sports bar')).
entail(prefer('alcohol'), require('establishment', 'bar&grill')).
entail(require('establishment', 'restaurant'), require('establishment', 'buffet')).

File: preference.py
from call_gpt import preference_classify
from utils import add_quote


def extract_prefer_preds(path):
    with open(path, 'r') as f:
        lines = f.readlines()
    
    classes = []
    for line in lines:
        if line.startswith('entail') and 'food type' in line:
            classes.append(line.split("'food type', ")[1].strip(').'))
    
    return classes


def get_prefer(preference_path, prefer_list):
    classes = extract_prefer_preds(preference_path)
    output = ''
    for p in prefer_list:
        if p not in classes:
            cuisines = preference_classify(', '.join(classes), p)
            cuisines = [x.strip() for x in cuisines.split(',')]
            cuisines = [x for x in cuisines if x]
            prefer = add_quote(p)
            for c in cuisines:
                cuisine = add_quote(c)
                pos_rule = "new_require('food type', " + cuisine + ") :- new_prefer(" + prefer + "), not not_require('food type', " + cuisine + ").\n"
                neg_rule = "new_not_require('food type', " + cuisine + ") :- new_not_prefer(" + prefer + "), not require('food type', " + cuisine + ").\n"
                ent_rule = "entail(prefer(" + prefer + "), require('food type', " + cuisine + ")).\n"

                output += pos_rule + neg_rule + ent_rule
    return output